File: 2022/haskell/app/Day2.hs
module Day2 (day2) where

import           System.IO

fp :: FilePath
fp = "../data/Day2.txt"

data Move = Rock | Paper | Scissors
    deriving (Eq)


-- Constants
win = 6 :: Int
tie = 3 :: Int
loss = 0 :: Int


-- Part 2 functions
toDraw :: Move -> Move
toDraw = id  -- toWin . toWin . toWin

toWin :: Move -> Move
toWin Rock = Paper
toWin Scissors = Rock
toWin Paper = Scissors

toLose :: Move -> Move
toLose = toWin . toWin


-- Part 1 functions
compareMoves :: Move -> Move -> Int
compareMoves left right 
  | left == right = tie
  | left == toWin right = win
  | otherwise = loss


decodeToMove :: String -> Move
decodeToMove "X" = Rock
decodeToMove "Y" = Paper
decodeToMove "Z" = Scissors
decodeToMove "A" = Rock
decodeToMove "B" = Paper
decodeToMove "C" = Scissors
decodeToMove s   = error $ "Broken key: " ++ s


decodeToFunction :: String -> (Move -> Move)
decodeToFunction "X" = toLose
decodeToFunction "Y" = toDraw
decodeToFunction "Z" = toWin
decodeToFunction s = error $ "Broken key" ++ s


weight :: Move -> Int
weight Rock     = 1
weight Paper    = 2
weight Scissors = 3


playRound1 :: String -> String -> Int
playRound1 theirs ours = weight ourMove + compareMoves ourMove theirMove
    where ourMove = decodeToMove ours
          theirMove = decodeToMove theirs


playRound2 :: String -> String -> Int
playRound2 theirs plan = weight ourMove + compareMoves ourMove theirMove
    where ourPlan = decodeToFunction plan
          theirMove = decodeToMove theirs
          ourMove = ourPlan theirMove


toRound1 :: [String] -> Int
toRound1 ws = if length ws >= 2
                then playRound1 (head ws) (ws !! 1)
                else error "Broken round"


toRound2 :: [String] -> Int
toRound2 ws = if length ws >= 2
                then playRound2 (head ws) (ws !! 1)
                else error "Broken round"


day2 :: IO ()
day2 = do
    contents <- readFile fp
    let part1 = sum $ map (toRound1 . words) $ lines contents
    let part2 = sum $ map (toRound2 . words) $ lines contents
    putStrLn $ "Part 1: " ++ show part1
    putStrLn $ "Part 2: " ++ show part2

File: 2022/rust/src/day2.rs
use std::str::FromStr;

static FILE_PATH: &'static str = "../data/Day2.txt";
static WIN      : u32 = 6;
static TIE      : u32 = 3;
static LOSS     : u32 = 0;

#[derive(Copy, Clone, PartialEq)]
enum Move {
    Rock, Paper, Scissors
}

impl Move {
    fn to_win(&self) -> Move {
        match self {
            Move::Rock => Move::Paper,
            Move::Paper => Move::Scissors,
            Move::Scissors => Move::Rock
        }
    }

    fn to_lose(&self) -> Move {
        self.to_win().to_win()
    }

    fn to_tie(&self) -> Move {
        *self
    }

    fn weight(&self) -> u32 {
        match self {
            Move::Rock => 1,
            Move::Paper => 2,
            Move::Scissors => 3
        }
    }

    fn compare(&self, other: Move) -> u32 {
        let result = if self == &other {
            TIE
        } else {
            if self == &other.to_win() {
                WIN
            } else {
                LOSS
            }
        };
        result + self.weight()
    }
}


impl FromStr for Move {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "A" => Ok(Move::Rock),
            "B" => Ok(Move::Paper),
            "C" => Ok(Move::Scissors),
            "X" => Ok(Move::Rock),
            "Y" => Ok(Move::Paper),
            "Z" => Ok(Move::Scissors),
            _   => Err(())
        }
    }
}


fn parse_str_to_function(text: &str) -> Result<fn(&Move) -> Move, &str> {
    match text {
        "X" => Ok(Move::to_lose),
        "Y" => Ok(Move::to_tie),
        "Z" => Ok(Move::to_win),
        _   => Err("Wrong key")
    }
}


pub fn day2() {
    let contents = std::fs::read_to_string(FILE_PATH).expect("Broken file");
    let lines: Vec<&str> = contents.split("\n").collect();
    let part1: u32 = lines.iter()
        .map(|x| {
            let row = x.split_once(" ");
            match row {
                Some((l, r)) => {
                    r.parse::<Move>()
                        .unwrap()
                        .compare(l.parse::<Move>().unwrap())
                },
                _ => 0
            }
        }).sum();
    let part2: u32 = lines.iter()
        .map(|x| {
            let row = x.split_once(" ");
            match row {
                Some((l, r)) => {
                    let their_move = l.parse::<Move>().unwrap();
                    parse_str_to_function(r).unwrap()(&their_move).compare(their_move)
                },
                _ => 0
            }
        }).sum();
    println!("Day 2:");
    println!("Part 1: {:?}", part1);
    println!("Part 2: {:?}", part2);
}
